ctrl_model_pinocchio_vs_franka:
  type: panda_torque_mpc/CtrlModelPinocchioVsFranka
  arm_id: $(arg arm_id)


ctrl_log_update_dt:
  type: panda_torque_mpc/CtrlLogUpdateDt
  arm_id: $(arg arm_id)


ctrl_playback_pd_plus:
    type: panda_torque_mpc/CtrlPlaybackPDplus
    arm_id: $(arg arm_id)
    joint_names:
        - $(arg arm_id)_joint1
        - $(arg arm_id)_joint2
        - $(arg arm_id)_joint3
        - $(arg arm_id)_joint4
        - $(arg arm_id)_joint5
        - $(arg arm_id)_joint6
        - $(arg arm_id)_joint7
    # traj_dir: "/home/mfourmy/Downloads/tsid_panda_traj_slow/"
    traj_dir: "/home/mfourmy/Downloads/croco_panda_traj_slow/"
    # traj_dir: "/home/mfourmy/Downloads/croco_panda_traj_mid/"
    # traj_dir: "/home/mfourmy/Downloads/croco_panda_traj_fast/"
    # traj_dir: "/home/mfourmy/Downloads/croco_panda_traj_fffast/"
    scale_ff: 1.0 
    Kp: 50 
    Kd: 5 
    publish_rate: 10.0
    alpha_dq_filter: 0.99 # should be close to 1 and <= 1.0
    saturate_dtau: true


ctrl_joint_space_ID:
    type: panda_torque_mpc/CtrlJointSpaceID
    arm_id: $(arg arm_id)
    joint_names:
        - $(arg arm_id)_joint1
        - $(arg arm_id)_joint2
        - $(arg arm_id)_joint3
        - $(arg arm_id)_joint4
        - $(arg arm_id)_joint5
        - $(arg arm_id)_joint6
        - $(arg arm_id)_joint7
    # Gains for IDControl
    Kp: 300  # Stiff
    Kd: 30  # Stiff
    # Kp: 100 
    # Kd: 10 
    # Kp: 50
    # Kd: 5
    # Gains for IDControlSimplified
    kp_gains:
        - 60.0
        - 60.0
        - 60.0
        - 60.0
        - 20.0
        - 10.0
        - 5.0
    kd_gains:
        - 5.0
        - 5.0
        - 5.0
        - 2.0
        - 2.0
        - 2.0
        - 1.0
    # Trajectory simulation
    delta_q:
        - -0.3
        -  0.3
        -  0.3
        -  0.3
        -  0.3
        -  0.3
        -  0.3
    period_q:
        - 5.0
        - 5.0
        - 5.0
        - 5.0
        - 5.0
        - 5.0
        - 5.0
    publish_rate: 10.0
    # 0: IDControl
    # 1: IDControlSimplified
    # 2: PDGravity
    # 3: PureGravity
    control_variant: 0
    use_pinocchio: true
    alpha_dq_filter: 0.99 # should be close to 1 and <= 1.0
    saturate_dtau: true


ctrl_task_space_ID:
    type: panda_torque_mpc/CtrlTaskSpaceID
    arm_id: $(arg arm_id)
    ee_frame_name: panda_hand
    joint_names:
        - $(arg arm_id)_joint1
        - $(arg arm_id)_joint2
        - $(arg arm_id)_joint3
        - $(arg arm_id)_joint4
        - $(arg arm_id)_joint5
        - $(arg arm_id)_joint6
        - $(arg arm_id)_joint7
    # Trajectory simulation
    delta_nu:
        - 0.30
        - 0.30
        - 0.30
        - 0.0
        - 0.0
        - 0.0
    period_nu:
        - 4.0
        - 4.0
        - 4.0
        - 5.0
        - 5.0
        - 5.0
    publish_rate: 10.0
    use_external_pose_publisher: false
    # 0: PosiPosture
    # 1: PosePosture  # TODO: not working -> pbe in jacobian
    # 2: TSIDPose
    control_variant: 2
    use_pinocchio: true
    alpha_dq_filter: 1.0 # should be close to 1 and <= 1.0
    # Controller configs
    # kp_ee: 500  # Stiff
    # kd_ee: 200 
    kp_ee: 100 
    kd_ee: 60 
    # kp_ee: 20
    # kd_ee: 5 
    kp_q: 100
    kd_q: 40
    w_ee: 1
    w_q: 0.001
    tau_limit_scale: 0.8
    v_limit_scale: 0.8
    # ee_task_mask: [1,1,1,1,1,1]
    ee_task_mask: [1,1,1,0,0,0]


ctrl_mpc_linearized:
    type: panda_torque_mpc/CtrlMpcLinearized
    arm_id: $(arg arm_id)
    ee_frame_name: panda_hand
    joint_names:
        - $(arg arm_id)_joint1
        - $(arg arm_id)_joint2
        - $(arg arm_id)_joint3
        - $(arg arm_id)_joint4
        - $(arg arm_id)_joint5
        - $(arg arm_id)_joint6
        - $(arg arm_id)_joint7

    publish_rate: 10.0
    alpha_dq_filter: 1.0 # should be close to 1 and <= 1.0

    ###############
    # JSID init controller params 
    Kp_jsid: 80 
    Kd_jsid: 10 
    dt_transition_jsid_to_mpc: 0.5

    use_riccati_gains: true

    ####################
    freq_solve: 100
    # OCP motion server configs
    # nb_shooting_nodes: 20

    dt_ocp: 0.05
    # nb_iterations_max: 1
    reference_is_placement: true   # otherwise, translation
    keep_original_ee_rotation: false
    # armature: [0.01,0.01,0.01,0.01,0.01,0.01,0.01]
    # armature: [0.001,0.001,0.001,0.001,0.001,0.001,0.001]
    armature: [0.,0.,0.,0.,0.,0.,0.0]
    
    #### PARAMETERS FROM NYC 
    nb_shooting_nodes: 15
    nb_iterations_max: 1
    solver_termination_tolerance: 0.0001              # Tolerance on KKT residual 
    max_qp_iter: 100    #125 
    qp_termination_tol_abs: 0.00001
    qp_termination_tol_rel: 0.00001

    ####################
    # frame pose/position
    # w_frame_running:  50
    # w_frame_terminal: 50
    w_frame_running:  10.0
    w_frame_terminal: 100.0

    # frame vel regularization
    w_frame_vel_running:  0.0
    w_frame_vel_terminal: 0.0
    diag_frame_vel: [
        1.,1.,1.,
        1.,1.,1.,
    ]

    # State regularization
    w_x_reg_running:  0.01
    w_x_reg_terminal: 0.01
    # High penalty on first joint -> tendancy to switch to 180 deg config
    diag_q_reg_running: [5., 1., 1., 1., 1., 1., 1.]
    diag_v_reg_running: [10., 10., 10., 10., 10., 10., 10. ]

    # Control regularization
    w_u_reg_running: 0.001
    diag_u_reg_running: [1.,1.,1.,1.,1.,1.,1.]

    # Obstacle avoidance
    collision_safety_margin: 0.20  # in meter 

    ##### WEIGHTS FROM THEO
    # targets params
    changing_weights: false
    pose_target1: [0.0, -0.35, 1.0, 0.0, 1.0, 0.0, 0.0]
    pose_target2: [0.0, 0.35, 1.0, 0.0, 1.0, 0.0, 0.0]
    publish_frequency: 0.25

    # Parameters to modify the curve of the target weight function
    w_slope: 13
    max_w: 1000 
    w_cut: 1
    #w_s=18.5 max_w=115, nb_nodes=40, niter=1 freq=0.5

    ################################
    # Visual servoing mode reference
    ################################
    
    # pose format: px, py, pz,    qx, qy, qz, qw
    pose_e_c: [0.137, -0.01, -0.059,  0.0869, -0.0866, -0.70365, 0.69986]  # Camera CALIBRATION d455 CIIRC
    # pose_e_c: [ 0.13620487, -0.02443776, -0.08476231,   -0.09459456, -0.08732901, 0.68741999, 0.71548848]  # Camera CALIBRATION d435 LAAS 20/12

    # Target relative pose
    # pose_c_o_ref: [0.029495, -0.01033, 0.758923,  0.99811875773742,    0.02157168338540,    -0.0190267711727,    0.05414415857087]   # Apriltag 0
    # pose_c_o_ref: [0., 0., 0.71,     1, 0, 0, 0 ]   # Apriltag 0
    # pose_c_o_ref: [-0.03017028, 0.0899488, 0.60250643,  0.44333751,  0.44584125, -0.53308101,  0.56626288] # cracker_box front tilted
    # pose_c_o_ref: [0., 0.1, 0.5,    -0.5, -0.5, 0.5, -0.5 ]   # cracker_box front 
    # pose_c_o_ref: [-0.0399200282198421, 0.06638413117011709, 0.4704659749834999, 
    #                0.1644332627969024, 0.6832111200594132, -0.5174592818349182, 0.48834781556460893]     # cracker_box side
    # pose_c_o_ref: [0., 0.0, 0.45,   -0.1474599121420375, -0.6722591020570695, 0.7138663759843439, -0.1292983806993129 ]   # mustard_bottle lookup up
    pose_c_o_ref: [-0.05, 0.02, 0.55, 0.7089126656918067, 0.1153152625077987, 0.14115075845828517, 0.6813381448142873]


    # pose_c_o_ref: [0.042518, -0.06673, 0.663273,   0.08617244980341546, 0.7010841883772583, -0.6874608345734711, 0.16867975762146092]    # mustard_bottle lookup up (LAAS)
    
    # pose_c_o_ref: [0.054, 0.0654, 0.5,   -0.1474599121420375, 0.40176469195778003, -0.21992470044436893, 0.6435384341970608 ]   # mustard_bottle rotated 4 oclock
    # pose_c_o_ref: [-0.017223091959642396, -0.00617516363648576, 0.252334571202528, 0.047980792158326166, -0.4866055057691095, 0.7345023199757922, -0.4705519453560494]  # mustard_bottle up-ish CLOSE
    # pose_c_o_ref: [-0.0005185746387237656, -0.019305824917591488, 0.1846942521241277, -0.12260664287122636, -0.6270232682029728, 0.7628840968106931, -0.09904932041158514]   # mustard_bottle up GRASP!
    # pose_c_o_ref: [-0.03013385739923637, 0.0555846759856413, 0.40886854260630845, 0.625720102455909, -0.5316429892823912, 0.4304552703857319, -0.37488436789089313]  # mug handle up
    # pose_c_o_ref: [0.062668356092972, 0.0726930998929625, 0.3809151875145119, 0.8983099359584542, 0.3853965546708258, -0.17030175493144784, -0.1245234473798262]  # large_clamp 8 hours
    # pose_c_o_ref: [0.04470781172080321, 0.0074622872137849496, 0.43623833556790576, -0.5689803061925258, 0.24554302353916055, 0.17620105506965178, 0.7648027731421333]  # tomato_soup_can front, looking down